ethereum_bridge/src/protocol/transactions/bridge_pool_roots.rs

```diff
@@ -0,0 +1,824 @@
+/// bridge pool root and nonce. Note that every signature
+/// passed into this function will be for the same
+/// root and nonce.
+///
+/// For roots + nonces which have been seen by a quorum of
+/// validators, the signature is made available for bridge
+/// pool proofs.
+pub fn apply_derived_tx<D, H>(
+    storage: &mut Storage<D, H>,
+    sigs: MultiSignedVext,
+) -> Result<TxResult>
+where
+    D: 'static + DB + for<'iter> DBIter<'iter> + Sync,
+    H: 'static + StorageHasher + Sync,
+{
+    if sigs.is_empty() {
+        return Ok(TxResult::default());
+    }
+    tracing::info!(
+        bp_root_sigs = sigs.len(),
+        "Applying state updates derived from signatures of the Ethereum \
+         bridge pool root and nonce."
+    );
+    let bp_root = parse_vexts(storage, sigs);
+    let voting_powers = utils::get_voting_powers(storage, &bp_root)?;
+
+    // apply updates to the bridge pool root.
+    let (mut changed, confirmed) = apply_update(
+        storage,
+        BridgePoolRoot(bp_root.root.clone()),
+        bp_root.seen_by.clone(),
+        &voting_powers,
+    )?;
+
+    // if the root is confirmed, update storage and add
+    // relevant key to changed.
+    if confirmed {
+        storage
+            .write(
+                &get_signed_root_key(),
+                bp_root
+                    .root
+                    .try_to_vec()
+                    .expect("Serializing a Keccak hash should not fail."),
+            )
+            .expect(
+                "Writing a signed bridge pool root to storage should not fail.",
+            );
+        changed.insert(get_signed_root_key());
+    }
+
+    // apply updates to the bridge pool nonce.
+    let (mut nonce_changed, confirmed) = apply_update(
+        storage,
+        BridgePoolNonce(bp_root.nonce.clone()),
+        bp_root.seen_by.clone(),
+        &voting_powers,
+    )?;
+    // add newly changed keys
+    changed.append(&mut nonce_changed);
+
+    // if the nonce is confirmed, update storage and add
+    // relevant key to changed.
+    if confirmed {
+        storage
+            .write(
+                &get_signed_nonce_key(),
+                bp_root
+                    .nonce
+                    .try_to_vec()
+                    .expect("Serializing a Uint should not fail."),
+            )
+            .expect("Writing a signed bridge pool nonce should not fail");
+        changed.insert(get_signed_nonce_key());
+    }
+    Ok(TxResult {
+        changed_keys: changed,
+        ..Default::default()
+    })
+}
+
+/// An Ethereum bridge pool root + nonce still awaiting
+/// a quorum of backing signatures to make in on chain.
+struct PendingQuorum {
+    /// The root of bridge pool being signed off on.
+    pub root: KeccakHash,
+    /// The nonce of bridge pool being signed off on.
+    pub nonce: Uint,
+    /// The validators who have already signed off
+    /// on this root + nonce
+    pub seen_by: Votes,
+}
```

Comment:
I think you don't need this type here, `Tally` already contains `seen_by`, and the body of the tally you write to storage could have `EthereumProof<(nonce, root)>` (this `EthereumProof` type is already defined in the `ethereum_bridge` crate). we do actually need to write the signatures to storage, as well, since later on we require them to relay transfers to Ethereum. in the `Tally`, if `tally.seen = false`, we know our proof is still considered incomplete (there isn't 2/3 stake backing it up)

Reply:
right, either way you can derive the new votes from `MultiSignedVext`. `PendingQuorum` here doesn't serve much purpose, IMO